sktime/classification/early_classification/_teaser.py

```diff
@@ -362,7 +371,7 @@ def _predict_proba(self, X, state_info=None):
         probas = np.array(
             [
                 probas[new_state_info[i][0] - (last_idx + 1)][i]
-                if accept_decision[i]
+                if accept_decision[i] and new_state_info[i][0] == last_idx + 1
```

Comment:
I think, this should be:


```
probas = np.array(
[
probas[new_state_info[i, 0] - (last_idx + 1)][i]
if accept_decision[i] and new_state_info[i, 0] <= next_idx
else [-1 for _ in range(self.n_classes_)]
for i in range(n_instances)
]
)
```

Reply:
I'm struggling to reproduce this actually, moved all the notebook code into a test, and it runs fine. Don't get any error raised in the notebook.

We don't have any probabilities to return below (last_idx + 1) I'm pretty sure, so we have to return [-1] for anything that has already made a decision.

Reply:
Hmm, ...

This test:
```
X = arrow_test_X[:, :, :100]
probas, decisions, states = teaser.predict_proba(X)
idx = np.where(np.array(probas) >= 0)
print(probas)
print(idx)
```

Returns always -1 for all probabilities. When you look into the decisions however, there are a couple of "true" decisions.

Reply:
This fix addresses a couple of things:

```
probas = np.array(
[
probas[new_state_info[i, 0] - (last_idx + 1)][i]
if accept_decision[i] and new_state_info[i, 0] <= next_idx
else [-1 for _ in range(self.n_classes_)]
for i in range(n_instances)
]
```

- Use `next_idx` instead of last+1. The test may do a couple of runs but not just one if the length is larger than the subsequent _classification_points
- use new_state_info[i, 0] instead of new_state_info[i][0] - pretty sure these are not the same for 2d data

Reply:
I think we are testing different things. The previous method broke a single run, I think this will break multiple consecutive runs. 

if we are on timestamp 5 and we already made a decision in the call at time stamp 3, we dont have any probabilities to return for the case that already made a decision.

```
        probas = np.array(
            [
                probas[new_state_info[i][0] - (last_idx + 1)][i]
                if accept_decision[i] and new_state_info[i][0] >= last_idx + 1
                else [-1 for _ in range(self.n_classes_)]
                for i in range(n_instances)
            ]
        )
```

changing to >= should work? for the test above at least.

Reply:
> if we are on timestamp 5 and we already made a decision in the call at time stamp 3, we dont have any probabilities to return for the case that already made a decision.

I do not understand.

I am on your notebook. You wrote an example how to use TEASER with partial data. And when you run this example (regardless of the length you give it), it will always return -1 as all probabilities. Even when the decisions were taken out. 

And the reason is: when calling predict_proba with a TS length, that is larger than the first classification point it will do multiple predictions (not just one) for all offsets that are smaller than the current length of the time series.

Reply:
But yet:

This
`                if accept_decision[i] and new_state_info[i, 0] <= next_idx`
seems to give the same result as:
`                if accept_decision[i] and new_state_info[i][0] >= last_idx + 1`

I personally find the first one more intuitive :)

Reply:
The second example in the notebook however needs a fix, too.

```
probas = np.array(
    372     [
    373         # probas[max(0, new_state_info[i, 0] - (last_idx + 1))][i]
--> 374         probas[new_state_info[i, 0] - (last_idx + 1)][i]
    375         if accept_decision[i] and new_state_info[i, 0] <= next_idx
    376         # if accept_decision[i] and new_state_info[i][0] >= last_idx + 1
    377         else [-1 for _ in range(self.n_classes_)]
    378         for i in range(n_instances)
    379     ]
    380 )
    382 return (probas, accept_decision, new_state_info)

IndexError: tuple index out of range
```

This might be fixed by:
`probas[max(0, new_state_info[i, 0] - (last_idx + 1))][i]`

Not sure, if this is the best fix?

Reply:
```
def test_bug():
    # Imports used in this notebook
    import numpy as np

    from sktime.classification.early_classification._teaser import TEASER
    from sktime.classification.interval_based import TimeSeriesForestClassifier
    from sktime.datasets import load_arrow_head

    # Load default train/test splits from sktime/datasets/data
    arrow_train_X, arrow_train_y = load_arrow_head(split="train", return_type="numpy3d")
    arrow_test_X, arrow_test_y = load_arrow_head(split="test", return_type="numpy3d")

    teaser = TEASER(
        random_state=0,
        classification_points=[25, 50, 75, 100, 125, 150, 175, 200, 251],
        estimator=TimeSeriesForestClassifier(n_estimators=10, random_state=0),
    )
    teaser.fit(arrow_train_X, arrow_train_y)

    X = arrow_test_X[:, :, :25]
    probas, decisions, states = teaser.predict_proba(X, state_info=None)

    print(probas)
    print(decisions)

    test_points = [50, 75, 100, 125, 150, 175, 200, 251]
    final_probas = np.zeros((arrow_test_X.shape[0], np.unique(arrow_test_y).shape[0]))
    final_decisions = np.zeros(arrow_test_X.shape[0])

    for i in test_points:
        X = arrow_test_X[:, :, :i]
        probas, decisions, states = teaser.predict_proba(X, state_info=states)

        print(probas)
        print(decisions)

        for n in range(arrow_test_X.shape[0]):
            if decisions[n] and final_decisions[n] == 0:
                final_probas[n] = probas[n]
                final_decisions[n] = i
```

This currently gives the output i would expect. When a decision becomes true for the first time, probabilities are returned, else -1.

that is with `if accept_decision[i] and new_state_info[i][0] >= last_idx + 1`

with `if accept_decision[i] and new_state_info[i, 0] <= next_idx` i get 

```
.0 = <range_iterator object at 0x00000155667F5C30>

        [
>           probas[new_state_info[i][0] - (last_idx + 1)][i]
            if accept_decision[i] and new_state_info[i, 0] <= next_idx
            else [-1 for _ in range(self.n_classes_)]
            for i in range(n_instances)
        ]
    )
E   IndexError: tuple index out of range
```

Reply:
because if we are doing them 1 by 1, we can only return probabilities for decisions made in that predict_proba call, not previous ones (old state_infos wont be updated, and will try to get probas that dont exist)

Reply:
Isn`t it possible with:

```
 probas = np.array(
            [
                probas[max(0, new_state_info[i, 0] - (last_idx + 1))][i]
                if accept_decision[i] and new_state_info[i, 0] <= next_idx
                else [-1 for _ in range(self.n_classes_)]
                for i in range(n_instances)
            ]
        )
```

i.e. if the probas aren't updated, we simply take the only one (first) entry?

Reply:
At least your test runs with that addition.

Reply:
The above code also works with the notebook with multiple calls.

Reply:
```
 probas = np.array(
            [
                probas[max(0, new_state_info[i, 0] - (last_idx + 1))][i]
                if accept_decision[i]
                else [-1 for _ in range(self.n_classes_)]
                for i in range(n_instances)
            ]
        )
```

dont think we need the and with that. Yeah should work, only difference is whether we want to return probabilities for decisions we have already made previously

Reply:
I would rather keep that information, otherwise we cannot compute earliness and accuracy easily without merging

Reply:
Are you ok, if I push your last code and the notebook?

Reply:
Yeah go ahead. 

```
        probas = np.array(
            [
                probas[
                    new_state_info[i][0] - (last_idx + 1)
                    if new_state_info[i][0] - (last_idx + 1) >= 0
                    else len(probas) - 1
                ][i]
                if accept_decision[i]
                else [-1 for _ in range(self.n_classes_)]
                for i in range(n_instances)
            ]
        )
```

This is uglier but will return the last set of probas instead of the first (in theory more accurate?) i'm not gonna push that it be used though :)

Reply:
Is there any difference?

Quickly tried the following with your test:

```
probas2 = np.array(
            [
                probas[max(0, new_state_info[i, 0] - (last_idx + 1))][i]
                if accept_decision[i]
                else [-1 for _ in range(self.n_classes_)]
                for i in range(n_instances)
            ]
        )


probas = np.array(
            [
                probas[
                    new_state_info[i, 0] - (last_idx + 1)
                    if new_state_info[i, 0] - (last_idx + 1) >= 0
                    else len(probas) - 1
                ][i]
                if accept_decision[i]
                else [-1 for _ in range(self.n_classes_)]
                for i in range(n_instances)
            ]
        )

print (probas != probas2)
```

Always false

Reply:
Only if you predict steps larger than 1 i.e. [10, 20, 30, 40, 50, 60] going from 10 to 30 then 50. It is pretty minor.

if you are going in steps of 1 then len(probas) - 1 always == 0, and if you are doing it all at once it doesn't really matter.

Reply:
I see. I am using 
```
        probas = np.array(
            [
                probas[max(len(probas) - 1, new_state_info[i, 0] - (last_idx + 1))][i]
                if accept_decision[i]
                else [-1 for _ in range(self.n_classes_)]
                for i in range(n_instances)
            ]
        )
```

then. This surely will not harm

Reply:
that will always return the last probability right?

Reply:
Just go with 0 if it is getting complicated, it doesn't really matter if we return the probabilities from time stamp 40 or 50 if we already returned a decision telling them to use timestamp 30.